app/Http/Controllers/v1/User/ContributionController.php

```diff
@@ -0,0 +1,67 @@
+<?php
+/**
+ * Manage the user contributions.
+ *
+ * @author Corentin Mercier <corentin@cmercier.fr>
+ *
+ * @copyright Copyright (c) 2018, SiMDE-UTC
+ * @license GNU GPL-3.0
+ */
+
+namespace App\Http\Controllers\v1\User;
+
+use App\Http\Controllers\v1\Controller;
+use Illuminate\Http\JsonResponse;
+use Illuminate\Http\Request;
+use App\Models\Semester;
+use App\Traits\Controller\v1\HasUsers;
+
+class ContributionController extends Controller
+{
+    use HasUsers;
+
+    /**
+     * Must be able to get user contributions.
+     */
+    public function __construct()
+    {
+        $this->middleware(
+            \Scopes::matchOneOfDeepestChildren('user-get-contributions'),
+            ['only' => ['index', 'show']]
+        );
+    }
+
+    /**
+     * List contributions of the current user.
+     *
+     * @param Request $request
+     * @param string  $user_id
+     * @return JsonResponse
+     */
+    public function index(Request $request, string $user_id=null)
+    {
+        $user = $this->getUser($request, $user_id);
+        $ginger = \Ginger::userByEmail($user->email);
+        $contributions = [];
+
+        foreach ($ginger->getContributions() as $contribution) {
+            $semesters = Semester::whereDate('begin_at', '<', $contribution->end_at)
```

Comment:
If i got 10 contributions, you will fetch 10 times your semesters. I think it is better that you fetch all semesters and then make your filter for each contribution